python/src/spark_rapids_ml/umap.py

```diff
@@ -0,0 +1,476 @@
+            )
+        else:
+            data_subset = dataset
+
+        input_num_workers = self.num_workers
+        # Force to single partition, single worker
+        self._num_workers = 1
+        if data_subset.rdd.getNumPartitions() != 1:
+            data_subset = data_subset.repartition(1)
+
+        pipelined_rdd = self._call_cuml_fit_func(
+            dataset=data_subset,
+            partially_collect=False,
+            paramMaps=None,
+        )
+        rows = pipelined_rdd.collect()
+        # Collect and concatenate row-by-row fit results
+        embeddings = [row["embedding_"] for row in rows]
+        raw_data = [row["raw_data_"] for row in rows]
+        del rows
+
+        model = UMAPModel(
+            embedding_=embeddings,
+            raw_data_=raw_data,
+            n_cols=len(raw_data[0]),
+            dtype=str(np.array(raw_data[0][0]).dtype),
+        )
+        model._num_workers = input_num_workers
+
+        self._copyValues(model)
+        self._copy_cuml_params(model)  # type: ignore
+
+        return model
+
+    def _fit_array_order(self) -> _ArrayOrder:
+        return "C"
+
+    def _get_cuml_fit_func(  # type: ignore
+        self, dataset: DataFrame
+    ) -> Callable[[FitInputType, Dict[str, Any]], Dict[str, Any],]:
+        """
+        This class overrides the parent function with a different return signature.
+        """
+        pass
+
+    def _get_cuml_fit_generator_func(
+        self,
+        dataset: DataFrame,
+        extra_params: Optional[List[Dict[str, Any]]] = None,
+    ) -> Callable[
+        [FitInputType, Dict[str, Any]], Generator[Dict[str, Any], None, None]
+    ]:
+        array_order = self._fit_array_order()
+
+        def _cuml_fit(
+            dfs: FitInputType,
+            params: Dict[str, Any],
+        ) -> Generator[Dict[str, Any], None, None]:
+            from cuml.manifold import UMAP as CumlUMAP
+
+            umap_object = CumlUMAP(
+                **params[param_alias.cuml_init],
+            )
+
+            df_list = [x for (x, _, _) in dfs]
+            if isinstance(df_list[0], pd.DataFrame):
+                concated = pd.concat(df_list)
+            else:
+                concated = _concat_and_free(df_list, order=array_order)
+
+            if dfs[0][1] is not None:
+                # If labels are provided, call supervised fit
+                label_list = [x for (_, x, _) in dfs]
+                if isinstance(label_list[0], pd.DataFrame):
+                    labels = pd.concat(label_list)
+                else:
+                    labels = _concat_and_free(label_list, order=array_order)
+                local_model = umap_object.fit(concated, y=labels)
+            else:
+                # Call unsupervised fit
+                local_model = umap_object.fit(concated)
+
+            embedding = local_model.embedding_
+            del local_model
+
+            for embedding, raw_data in zip(embedding, concated):
```

Comment:
Is this not too slow to create a DataFrame for each row?   Does working in chunks and yielding List[List[]] for each chunk in each corresponding "Row" work?